admin-auth/src/test/kotlin/com/epam/drill/admin/auth/OAuthModuleTest.kt

```diff
@@ -0,0 +1,245 @@
+import org.mockito.kotlin.whenever
+import java.security.KeyPair
+import java.security.KeyPairGenerator
+import java.security.interfaces.RSAPrivateKey
+import java.security.interfaces.RSAPublicKey
+import java.net.URL
+import kotlin.test.*
+
+typealias RequestHandler = Pair<String, suspend MockRequestHandleScope.(HttpRequestData) -> HttpResponseData>
+
+class OAuthModuleTest {
+
+    private val testOAuthServerHost = "some-oauth-server.com"
+    private val testDrillHost = "example.com"
+    private val testClientId = "test-client"
+    private val testClientSecret = "test-secret"
+    private val keyPair = generateRSAKeyPair()
+    private val rsa256 = Algorithm.RSA256(keyPair.public as RSAPublicKey, keyPair.private as RSAPrivateKey)
+
+    @Mock
+    lateinit var mockJwkProvider: JwkProvider
+    @Mock
+    lateinit var jwk: Jwk
+
+    @BeforeTest
+    fun setup() {
+        MockitoAnnotations.openMocks(this)
+    }
+
+    @Test
+    fun `given valid oauth2 token, jwt protected request must succeed`() {
+        val testUsername = "test-user"
+        val testJwkKey = "test-jwk"
+        val testIssuer = "test-issuer"
+
+        withTestApplication({
+            environment {
+                put("drill.auth.oauth2.issuer", testIssuer)
+            }
+            withTestOAuthModule {
+                mockJwkProvider()
+            }
+            routing {
+                authenticate("jwt") {
+                    get("/protected") {
+                        call.respond(HttpStatusCode.OK)
+                    }
+                }
+            }
+        }) {
+            whenever(mockJwkProvider.get(testJwkKey)).thenReturn(jwk)
+            whenever(jwk.publicKey).thenReturn(keyPair.public)
+            whenever(jwk.algorithm).thenReturn("RS256")
+
+            with(handleRequest(HttpMethod.Get, "/protected") {
+                addJwtToken(
+                    username = testUsername,
+                    issuer = testIssuer,
+                    algorithm = rsa256
+                ) {
+                    withHeader(mapOf("kid" to testJwkKey))
+                    withClaim("preferred_username", testUsername)
+                    withClaim(
+                        "realm_access",
+                        mapOf("roles" to listOf("USER"))
+                    ) //TODO remove after adding claim mapping tests
+                }
+            }) {
+                assertEquals(HttpStatusCode.OK, response.status())
+            }
+        }
+    }
+
+    @Test
+    fun `oauth login request must be redirected to oauth2 authorize url`() {
+        withTestApplication({
+            withOAuthTestEnvironment()
+            withTestOAuthModule()
+        }) {
+            with(handleRequest(HttpMethod.Get, "/oauth/login")) {
+                assertEquals(HttpStatusCode.Found, response.status())
+
+                val url = URL(response.headers[HttpHeaders.Location])
+                val queryParams = url.query?.split("&")?.associate {
```

Comment:
move parsing code to `parseQueryParams` function (either lambda right here or move it to the test utils) to improve readability